src/shogun/mathematics/linalg/Vector.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2016, Shogun-Toolbox e.V. <shogun-team@shogun-toolbox.org>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  1. Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: 2016 Pan Deng, Soumyajit De, Viktor Gal
+*/
+
+#include <shogun/lib/config.h>
+#include <shogun/lib/SGVector.h>
+#include <memory>
+
+#ifndef LINALG_VECTOR_H__
+#define LINALG_VECTOR_H__
+
+#ifdef HAVE_CXX11
+
+namespace shogun
+{
+
+/** Vector structure
+ * for both CPU and GPU vector
+ */
+template <class T>
+class Vector
+{
+
+	typedef typename std::aligned_storage<sizeof(T), alignof(T)>::type aligned_t;
+
+public:
+	/** Default Constructor */
+	Vector();
+
+	Vector(SGVector<T> const &vector);
+
+	/** Copy Constructor */
+	Vector(Vector<T> const &vector);
+
+	~Vector();
+
+	Vector& operator=(SGVector<T> const &vector);
+
+	Vector& operator=(Vector<T> const &vector);
+
+	operator SGVector<T>();
+
+	/** Data Storage
+	 * @return whether data is on GPU
+	*/
+	bool onGPU() const;
+
+	T* data();
+
+	T const * data() const;
+
+	index_t size() const;
+
+	void transferToGPU();
+
+	void transferToCPU();
+
+	/** Operator overload for vector read only access
+	 *
+	 * @param index dimension to access
+	 *
+	 */
+	inline const T& operator[](index_t index) const
+	{
+		return m_data.get()[index];
+	}
+
+	/** Operator overload for vector read only access
+	 *
+	 * @param index dimension to access
+	 *
+	 */
+	inline T& operator[](index_t index)
+	{
+		return m_data.get()[index];
```

Comment:
This should return the data from GPU memory if onGPU is true :)
Maybe move these to cpp? Otherwise you'll have to include GPUVectorImpl here which is something we want to avoid.